cmd/lakectl/cmd/fs_cp.go

```diff
@@ -0,0 +1,145 @@
+
+		if sourceURI.Repository != destinationURI.Repository {
+			Die("Can only copy files in the same repo", 1)
+		}
+
+		client := getClient()
+
+		ctx := cmd.Context()
+
+		if !recursive {
+			err := copyObject(ctx, client, sourceURI, destinationURI)
+			if err != nil {
+				DieErr(err)
+			}
+			return
+		}
+
+		success := true
+		var errorsWg sync.WaitGroup
+		errors := make(chan error)
+		errorsWg.Add(1)
+		go func() {
+			defer errorsWg.Done()
+			for err := range errors {
+				_, _ = fmt.Fprintln(os.Stderr, err)
+				success = false
+			}
+		}()
+
+		var copyWg sync.WaitGroup
+		paths := make(chan string)
+		copyWg.Add(concurrency)
+
+		for i := 0; i < concurrency; i++ {
+			go copyObjectWorker(ctx, client, sourceURI, destinationURI, paths, errors, &copyWg)
+		}
+
+		prefix := *sourceURI.Path
+		var paramsDelimiter apigen.PaginationDelimiter = ""
+		var from string
+		pfx := apigen.PaginationPrefix(prefix)
+		for {
+			params := &apigen.ListObjectsParams{
+				Prefix:    &pfx,
+				After:     apiutil.Ptr(apigen.PaginationAfter(from)),
+				Delimiter: &paramsDelimiter,
+			}
+			resp, err := client.ListObjectsWithResponse(cmd.Context(), sourceURI.Repository, sourceURI.Ref, params)
+			DieOnErrorOrUnexpectedStatusCode(resp, err, http.StatusOK)
+			if resp.JSON200 == nil {
+				Die("Bad response from server", 1)
+			}
+
+			results := resp.JSON200.Results
+			for i := range results {
+				paths <- results[i].Path
+			}
+
+			pagination := resp.JSON200.Pagination
+			if !pagination.HasMore {
+				break
+			}
+			from = pagination.NextOffset
+		}
+
+		close(paths)
+		copyWg.Wait()
+		close(errors)
+		errorsWg.Wait()
+		if !success {
+			os.Exit(1)
+		}
+	},
+}
+
+func copyObjectWorker(ctx context.Context, client apigen.ClientWithResponsesInterface, sourceURI, destinationURI *uri.URI, paths <-chan string, errors chan<- error, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	srcPrefix := strings.TrimSuffix(*sourceURI.Path, "/")
+	dstPrefix := strings.TrimSuffix(*destinationURI.Path, "/")
+
+	for srcObjPath := range paths {
+		rel := strings.TrimPrefix(srcObjPath, srcPrefix+"/")
+		destObjPath := dstPrefix
+		if rel != "" {
+			destObjPath = path.Join(dstPrefix, rel)
+		}
+
+		resp, err := client.CopyObjectWithResponse(ctx, sourceURI.Repository, sourceURI.Ref, &apigen.CopyObjectParams{
+			DestPath: destObjPath,
+		},
+			apigen.CopyObjectJSONRequestBody{
+				SrcPath: srcObjPath,
+				SrcRef:  &sourceURI.Ref,
+			},
+		)
+		if err = RetrieveError(resp, err); err != nil {
+			errors <- fmt.Errorf("cp %q -> %q: %w", srcObjPath, destObjPath, err)
+		}
+	}
+}
+
+func copyObject(ctx context.Context, client apigen.ClientWithResponsesInterface, sourceURI, destinationURI *uri.URI) error {
+	resp, err := client.CopyObjectWithResponse(ctx, sourceURI.Repository, sourceURI.Ref, &apigen.CopyObjectParams{
+		DestPath: *destinationURI.Path,
+	}, apigen.CopyObjectJSONRequestBody{
+		SrcPath: *sourceURI.Path,
+		SrcRef:  &sourceURI.Ref,
+	})
```

Comment:
I do not understand how this copies from a ref to another branch: destinationURI.Ref is ignored.  I would expect e.g.
```suggestion
	resp, err := client.CopyObjectWithResponse(ctx, destinationURI.Repository, destinationURI.Ref, &apigen.CopyObjectParams{
		DestPath: *destinationURI.Path,
	}, apigen.CopyObjectJSONRequestBody{
		SrcPath: *sourceURI.Path,
		SrcRef:  &sourceURI.Ref,
	})
```
(but I could be wrong, it's been a long while...).